File: src/main/java/org/kouv/cornea/data/Attributes.java
package org.kouv.cornea.data;

import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.Nullable;

import java.util.IdentityHashMap;
import java.util.Map;
import java.util.Objects;

public final class Attributes {
    private final Map<AttributeKey<?>, Object> map = new IdentityHashMap<>();

    @ApiStatus.Internal
    public Attributes() {
    }

    public boolean contains(AttributeKey<?> key) {
        Objects.requireNonNull(key);
        return map.containsKey(key);
    }

    @SuppressWarnings("unchecked")
    public <T> @Nullable T get(AttributeKey<T> key) {
        Objects.requireNonNull(key);
        return (T) map.get(key);
    }

    public <T> void set(AttributeKey<T> key, T value) {
        Objects.requireNonNull(key);
        Objects.requireNonNull(value);
        map.put(key, value);
    }

    public <T> void remove(AttributeKey<T> key) {
        Objects.requireNonNull(key);
        map.remove(key);
    }
}

File: src/main/kotlin/org/kouv/cornea/data/Attributes.kt
package org.kouv.cornea.data

import java.util.*

public class AttributeKey<T>(
    public val name: String
)

public class Attributes internal constructor() {
    private val map: MutableMap<AttributeKey<*>, Any> = IdentityHashMap()

    public operator fun contains(key: AttributeKey<*>): Boolean {
        return key in map
    }

    @Suppress("UNCHECKED_CAST")
    public operator fun <T : Any> get(key: AttributeKey<T>): T? {
        return map[key] as? T
    }

    public operator fun <T : Any> set(key: AttributeKey<T>, value: T) {
        map[key] = value
    }

    public fun <T : Any> remove(key: AttributeKey<T>) {
        map.remove(key)
    }
}